{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "99861961",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import csv\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "class readCSV:\n",
    "    def __init__(self, filename= None):\n",
    "        self.filename = filename\n",
    "    \n",
    "    def csvreader(self):\n",
    "        csvdata = [] \n",
    "        with open(self.filename,'r') as file:\n",
    "            csvreader = csv.reader(file)\n",
    "            header = next(csvreader)\n",
    "            for row in csvreader:\n",
    "                csvdata.append(row)\n",
    "        return csvdata        \n",
    "    \n",
    "    def exonparam(self,Ecsvdata):\n",
    "        \n",
    "        pos_exon = {}\n",
    "        neg_exon = {}\n",
    "        \n",
    "        for element in Ecsvdata:\n",
    "            rangelist = []\n",
    "            if element[1] == '+':\n",
    "                startlist = element[3].split(',')\n",
    "                startlist.remove('')\n",
    "\n",
    "                stoplist = element[4].split(',')\n",
    "                stoplist.remove('')\n",
    "                \n",
    "                for position in range(0,len(startlist)):\n",
    "                    setinit = range(int(startlist[position]),int(stoplist[position]))\n",
    "                    rangelist.append(setinit)\n",
    "                \n",
    "                pos_exon.update({element[0]:rangelist})\n",
    "\n",
    "                    \n",
    "            elif element[1] == '-':\n",
    "                startlist = element[3].split(',')\n",
    "                startlist.remove('')\n",
    "\n",
    "                stoplist = element[4].split(',')\n",
    "                stoplist.remove('')\n",
    "                \n",
    "                for position in range(0,len(startlist)):\n",
    "                    setinit = range(int(startlist[position]),int(stoplist[position]))\n",
    "                    rangelist.append(setinit)\n",
    "                    neg_exon.update({element[0]:rangelist})\n",
    "            \n",
    "                    \n",
    "        return pos_exon, neg_exon\n",
    "    \n",
    "    \n",
    "    def repeatparam(self, Rcsvdata):\n",
    "        \n",
    "        repeats = []\n",
    "        \n",
    "        for element in Rcsvdata:\n",
    "            \n",
    "            #rangelist = []\n",
    "            \n",
    "            if element[1] == '+':\n",
    "                setinit = range(int(element[2]),int(element[3]))\n",
    "                repeats.append([element[0],'+',setinit,element[2],element[3]])\n",
    "            \n",
    "            elif element[1] == '-':\n",
    "                setinit = range(int(element[2]),int(element[3]))\n",
    "                repeats.append([element[0],'-',setinit,element[2],element[3]])\n",
    "        \n",
    "        return repeats\n",
    "    \n",
    "    def overlap(self,dict_of_ranges,repeat_data):\n",
    "        overlapdata = []\n",
    "        for isoform in dict_of_ranges:\n",
    "            for gene_id, values in isoform.items():\n",
    "                \n",
    "                for exon in values:\n",
    "                    exon = set(exon)\n",
    "                \n",
    "                    for repeat in repeat_data:\n",
    "                        set_of_ranges = set(repeat[2])\n",
    "            \n",
    "                        #for repeat,ranges in repeat.items():\n",
    "                            #set_of_ranges = set(ranges)\n",
    "                            #print(set_of_ranges)\n",
    "                            \n",
    "                            \n",
    "                        if len(exon.intersection(set_of_ranges)) >= 1:\n",
    "                            if len(set_of_ranges) == len(exon.intersection(set_of_ranges)):\n",
    "                                overlapdata.append([gene_id,repeat[0],repeat[1],repeat[3],repeat[4],'exon',len(set_of_ranges)])\n",
    "                                    \n",
    "                                    \n",
    "                            elif len(set_of_ranges) != len(exon.intersection(set_of_ranges)):\n",
    "                                    \n",
    "                                overlapdata.append([gene_id,repeat[0],repeat[1],repeat[3],repeat[4],'junction',len(exon.intersection(set_of_ranges))])\n",
    "                                    \n",
    "                                    \n",
    "                            \n",
    "                        elif len(exon.intersection(set_of_ranges)) == 0:\n",
    "                                \n",
    "                            overlapdata.append([gene_id,repeat[0],repeat[1],repeat[3],repeat[4],'intron', 0])\n",
    "        return overlapdata\n",
    "                        \n",
    "        \n",
    "\n",
    "##########################################################################################   \n",
    "    \n",
    "def main(filename = None):\n",
    "    \n",
    "    ExonReader = readCSV('KRAS.csv')\n",
    "    Exon_Raw_Data = ExonReader.csvreader()\n",
    "    \n",
    "    RepeatsReader = readCSV('KRAS_repeats.csv')\n",
    "    Repeats_Raw_Data = RepeatsReader.csvreader()\n",
    "    \n",
    "    Ecsvdata = []\n",
    "    for element in Exon_Raw_Data:\n",
    "        Ecsvdata.append([element[0],element[2],element[7], element[8], element[9]])\n",
    "    Rcsvdata = []   \n",
    "    for element in Repeats_Raw_Data:\n",
    "        Rcsvdata.append([element[10],element[9],element[6],element[7]])\n",
    "    \n",
    "    e_rangedata = ExonReader.exonparam(Ecsvdata)\n",
    "    \n",
    "    r_rangedata = RepeatsReader.repeatparam(Rcsvdata)\n",
    "    \n",
    "    #intersectiondata = isoform,repeat_name,repeat_strand,start,stop,classification,intersection \n",
    "    intersectiondata = RepeatsReader.overlap(e_rangedata,r_rangedata)\n",
    "    \n",
    "    \n",
    "    #gene, isoform, MER5B_range=chr1:11677-11780_strand=-, chrom, start, stop, instrand, genstrand, classification, per overlap\n",
    "    \n",
    "    finaldata = []\n",
    "    \n",
    "    for index in range(0,len(intersectiondata)):\n",
    "        intersectiondata[index].append(Exon_Raw_Data[0][1])\n",
    "        intersectiondata[index].append(Exon_Raw_Data[0][2])\n",
    "    \n",
    "    \n",
    "    for element in intersectiondata:\n",
    "        newlist = [element[0], element[1]+'_range='+element[7]+':'+element[3]+'-'+element[4]+'_strand='+element[2],element[7],element[3],element[4],element[2],element[8],element[5],element[6]]\n",
    "        finaldata.append(newlist)\n",
    "    \n",
    "    overlaps = pd.DataFrame(finaldata, columns= ['isoform','insertion_name','chrom','start','stop','instrand','genstrand','classification','overlap_count'])\n",
    "    overlaps.to_csv('TE_Overlap.csv', index=False)\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "    \n",
    "    \n",
    "## make it robust, testing figure out how to build tests\n",
    "## Send mySEQL code, automate \n",
    "## make this usable, command line option integration\n",
    "## LEARN HOW TO USE GITHUB\n",
    "## Make more modular classes \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53156482",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
